README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
endpointR <img src="https://talbotsr.com/endpointR/logo.png" align="right" height="139" />
==========================================================================================

endpointR
=========

<!-- badges: start -->
[![Travis build status](https://travis-ci.org/mytalbot/endpointR.svg?branch=master)](https://travis-ci.org/r-lib/usethis) <!-- badges: end -->

The endpointR is a computational tool for aiding scientists in laboratory animal research to identify humane endpoints. Originally it was developed for body weight (change) data but it can be used with any (longitudinal/time series) data. However, it is not meant as substitution for experience and reason - but it may help in any decision making process concerning the wellbeing of an animal.

Installation
------------

You can install the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("mytalbot/endpointR")
library(endpointR)
```

endpointR Shiny App
-------------------

The endpointR is ready-to-use as a Shiny app for body weight data analysis under the following link.

[endpointR](https://calliope.shinyapps.io/endpointer/)

Please restrict your online time on the app and close the browser when finished as we're only using a public license. Thank you!

Example
-------

This is a basic example uses the internal Glioma data from Helgers & Talbot et al 2019 (under review) to indicate potential endpoints for individual animals. The input variables are raw body weight data.

``` r
library(endpointR)
a      <- 1
td     <- ep_select(gliodat, a) 
danger <- epR(td        = td,
              org       = FALSE,
              wl        = 6,
              SDwdth    = 2,
              mad       = FALSE,
              blind     = TRUE)
danger
#>    n timepoint value where
#> 1 15        12 99.25 lower
#> 2 15        15 94.14 lower
```


NEWS.md
# endpointR 0.0.1

* Added a `NEWS.md` file to track changes to the package.


scripts/epR_svenja_bw_test.R


library(endpointR)
library(RELSA)

# Load Laura Act Data -----------------------------------------------------
raw       <- relsa_load("C:/MHH Bleich/Papers/PostOPPaper/Data/svenja_dss_mod.txt" )
raw       <- raw[, c(1, 2, 4, 5)]

# Chose animal ------------------------------------------------------------
a         <- 31
dat       <- raw[raw$id==unique(raw$id)[a],  c(1,3,4)]
td        <- reshape(dat, idvar = "id", timevar = "day", direction = "wide")
td        <- as.numeric(td[2:length(td)])

result    <- epR(td        = td,
                 org       = T,
                 wl        = 2,
                 SDwdth    = 2,
                 mad       = T,
                 cex       = 1.4,
                 cex.lab   = 1.2,
                 blind     = F,
                 ignupr    = TRUE,
                 ylim      = c(70,120), #c(300,350),
                 xlim      = c(0,20),
                 xlab      = "day")
print(result)


td        = td
org       = T
wl        = 2
SDwdth    = 2
mad       = F
cex       = 1.4
cex.lab   = 1.2
blind     = F
ignupr    = TRUE
ylim      = c(70,120)
xlim      = c(0,20)
xlab      = "day"



R/ep_load.R
#' Loads Data into the endpointeR
#'
#' Data from a *.txt file will be read into the function so that it can be fed to the endpointR.
#' Longitudinal data need to be in the wide format. They must have an index column with animal id,
#' a descriptive column such as treatment. The following columns contain values for consecutive time points. There should be
#' no missing values until the last measured data point.
#'
#' @param file path and filename as a string (e.g. C:/testfolder/testdata.txt)
#'
#' @importFrom utils read.table
#'
#' @return data.frame
#'
#' @export
#'
ep_load      <- function(file){

  testdata   <- data.frame(read.table(file, header = TRUE, sep = "\t") )

  animals    <- length(unique(testdata[,1]))

  # print(paste("Success! Data loaded with n = ", animals, sep=""))

  return(testdata)
}


R/gliodat.R
#' Rat Body Weight Data Glioblastoma Model
#'
#' Rat body weight data from a intracranial Glioblastoma model. 63 unique animals in three treatment groups (Control,
#' Microinjection and Resection) were monitored on a daily basis. The general model indicates that animal well-being is not substantially
#' impaired until a certain time-point at around 12-16 days is reached. Here, most animals decompensate rapidly and had to be
#' euthanized because the defined humane endpoint for the study was reached. The maximum follow-up is 28 days, although most animals had to
#' be euthanized well before that. In the resection group day 8 is the day of operation. These animals survive longer but a small drop in
#' body weight can be seen after day 8, together with rapid recovery.
#'
#' @docType data
#'
#' @usage data(gliodat)
#'
#' @format A 63 x 31 data frame with post-op data.
#'
#' @keywords datasets
#'
#' @examples
#' data(gliodat)
#' head(gliodat)
"gliodat"


R/endpoint.R
#' Endpointer function
#'
#' The \code{epR} function is a tool for the identification of humane endpoints using single outcome variables
#' from laboratory animal experiments. Originally, it was developed for using body weight but as values are normalized
#' other continous variables are suited as well. The algorithm is highly functional in identifying strong deviations from
#' a windowed normality. The hypothesis behind this is: larger deviations always point to severity.
#'
#' @param td testdata data.frame with n unique rows and p subsequent time points (e.g. days)
#' @param org boolean (TRUE/FALSE) for using original values. If FALSE, data are normalized.
#' @param wl SD window length (default is 6)
#' @param SDwdth width of the standard deviation around the moving average (default is 2.5)
#' @param mad boolean - use mean absolute deviation as quasi-clinical scoring constraint (default FALSE)
#' @param ltype line type in shown plot
#' @param dotcolor color of the shown dots (default "black")
#' @param uprcol color of the upper threshold violation indicators
#' @param lwrcol color of the upper threshold violation indicators
#' @param cex point size
#' @param cex.axis axis tick size
#' @param cex.lab label size
#' @param xlim range of x-axis (if set to NULL (default), plot will adapt automatically to given range - may not be nice!)
#' @param ylim range of <-axis (if set to NULL (default), plot will adapt automatically to given range - may not be nice!)
#' @param pch R-specific plot symbol for shown dots (default is 19)
#' @param blind boolean (TRUE/FALSE) - if set to TRUE, no plot will be shown (default is FALSE)
#' @param ignupr boolean (TRUE/FALSE) - ignore upper threshold violations (default is FALSE)
#' @param xlab x-axis label
#' @param ylab y-axis label
#' @param main title
#'
#' @importFrom stats sd complete.cases
#' @import graphics
#'
#' @return data.frame with enpointeR results (n=number of data points, timepoint=index of marked endpoint,
#' where=upper or lower boundary)
#'
#' @export
#'
#' @examples
#' epR(as.numeric(gliodat[1,3:length(gliodat[1,])]), blind=TRUE )
#'
epR <- function(td = td, org=FALSE, wl = 6, SDwdth = 2, mad=FALSE, ltype = "b", dotcolor = "black", uprcol = "darkgreen",
                lwrcol = "magenta", cex = 1, cex.axis = 1, cex.lab = 1, xlim = NULL, ylim = NULL, pch = 19, blind = FALSE,
                ignupr = FALSE, xlab = "time", ylab = "Moving average (%)", main = NULL){

  # essentials --------------------------------------------------------------
  pointlength    <- length(td)
  xachse         <- seq(1,pointlength)

  if(org == TRUE){
    W            <- td
  }else{
    epdat        <- (td / td[1])*100

  # calculate windowed moving average ---------------------------------------
    W <- c()
    S <- c()
    for (i in wl:pointlength ){
      floeating.mean <- mean(epdat[i:(i-(wl-1))])
      w              <- (epdat[i] / floeating.mean) * 100
      W[i]           <- w
    }
    W[1:wl]<- rep(100,wl)
  }

  # calculate backwards windowed SD and mean --------------------------------
  mysd   <- c()
  mymean <- c()
  for (i in pointlength:wl){
    mySD     <- sd(W[(i-1):(i-wl)],   na.rm=TRUE)
    mysd[i]  <- mySD
    myMean   <- mean(W[(i-1):(i-wl)], na.rm=TRUE)
    mymean[i]<- myMean
  }
  mysd[mysd==0] <- NA # remove 0, otherwise first point gets X-ed
  mysd       <-  mysd*SDwdth

  # calculate MAD -----------------------------------------------------------
  if(mad==TRUE){
  myMAD <- c()
  for (i in wl:pointlength ){
    myMAD[i] <- mean(W[(i-1):((i-1)-(wl-1))]) - W[i]
  }
  myMAD      <- round(myMAD,2)

  # constrain SD by MAD -----------------------------------------------------
  madscore   <- ceiling(abs(myMAD))
  madscore[madscore == 0] <-1
  mysd       <- mysd / madscore
  }else{}

  # plotting ----------------------------------------------------------------
  if(blind==FALSE){
    plot(xachse,
         W,
         type     = ltype,
         pch      = pch,
         col      = dotcolor,
         xlim     = xlim,
         ylim     = ylim,
         cex.lab  = cex.lab,
         cex.axis = cex.axis,
         cex      = cex,
         xlab     = xlab,
         ylab     = ylab,
         main     = main)

    lines(xachse, mymean + mysd, col="blue", lty = 2, cex = .5, lwd = 2.5)
    lines(xachse, mymean - mysd, col="blue", lty = 2, cex = .5, lwd = 2.5)

    abline(h = 100, col = "black",  lwd = 2, lty = 3)
    abline(v = wl , col = "gray40", lwd = 1, lty = 3)



  }else{}

  # diagnostics -------------------------------------------------------------
  result <- NULL
  if(ignupr==TRUE){
    lower            <- which(W < mymean - mysd)
    lower.time.idx   <- if(length(lower) == 0) NA else lower

    result           <- data.frame(n = length(W), timepoint = lower.time.idx)
    result$where     <- rep("lower", length(lower.time.idx))

    if(blind==FALSE){
      points(which(W< mymean-mysd), W[which(W< mymean-mysd)], col = lwrcol, pch = 4, cex = 3, lwd = 3)
    }else{}

  }else{
    lower            <- which(W < mymean - mysd)
    lower.time.idx   <- if(length(lower) == 0) NA else lower

    # capture errors if there are no lower threshold violations
    if(all(is.na(lower.time.idx)) != TRUE){
      result         <- data.frame(n = length(W), timepoint = lower.time.idx, value = round(W[lower.time.idx],2))
      result         <- result[order(result$timepoint, decreasing = F),]
    }else{
      result         <- data.frame(n = length(W), timepoint = NA, value =NA)
      result         <- result[order(result$timepoint, decreasing = F),]
    }

    upr              <- which(W > mymean + mysd)
    upr.time.idx     <- if(length(upr) == 0) NA else upr

    result           <- rbind(result, data.frame(n = length(W), timepoint = upr.time.idx,
                                                 value = if(all(is.na(W[upr.time.idx])) == TRUE) NA else round(W[upr.time.idx],2) ))
    result$where     <- append(rep("lower", length(lower.time.idx)), rep("upper", length(upr.time.idx)) )

    if(blind==FALSE){
      points(which(W< mymean-mysd), W[which(W< mymean-mysd)], col = lwrcol, pch = 4, cex = 3, lwd = 3)
      points(which(W> mymean+mysd), W[which(W> mymean+mysd)], col = uprcol, pch = 4, cex = 3, lwd = 3)
    }else{}

  }

  if(all(is.na(result$timepoint)) == TRUE){
    result           <- data.frame(n = length(W), timepoint = NA, where = NA)
  }else{
    result           <- result[complete.cases(result), ]
  }

  return(result)
}





R/ep_select.R
#' Animal Selection for endpointR Analysis
#'
#' Using the testdata output from \code{ep_load}, the \code{ep-select} function subsets the testdata by a
#' user-defined unique animal index. If the input is numeric, e.g. n=1 the first animal in testdata will be chosen. If the
#' input is a character string, e.g. "ZW21", this specific animal from the testdata will be chosen. Animal ids should not be numeric.
#'
#' @param testdata data.frame with test data as provided by \code{ep_load}
#' @param n animal index (numeric or character (animal id))
#'
#' @return vector of a single animal for distinct time points.
#'
#' @export
#'
#' @examples
#' ep_select(gliodat, n = 1)
#'
ep_select      <- function(testdata, n = 1){

  if(is.numeric(n) == TRUE){
    test       <- testdata[n,3:length(testdata)]
    orgw       <- as.numeric(test)
    orgw       <- orgw[!is.na(orgw)]
  }else{
    test       <- testdata[testdata[,1] == n,]
    orgw       <- as.numeric(test)
    orgw       <- orgw[!is.na(orgw)]
    n          <- which(testdata[,1] == n)
  }

  # print( data.frame(testdata[n, c(1,2) ], datapoints = length(orgw)) )

  return(orgw)
}


scripts/epR_perf.R

library(endpointR)

tp         <- c()
fp         <- c()
no_fa      <- c()
SP         <- NULL
for(i in 1:63){
  SP       <- NULL
  SP       <- epR(as.numeric(ep_select(gliodat, i)), SDwdth = 1, blind = TRUE )

  # if default... -----------------------------------------------------------
  default    <- ifelse((length(SP) == 0) == TRUE,1,0)
  if(default == 1){
    tp[i]    <- NA
    fp[i]    <- NA
    no_fa[i] <- NA
  }else{
    tp[i]    <- ifelse(sum(SP$timepoint       == unique(SP$n), na.rm = TRUE) == 1, 1, 0)
    fp[i]    <- ifelse( (length(SP$timepoint) != unique(SP$n)) == TRUE, 1, 0)
    no_fa[i] <- sum(SP$timepoint              != unique(SP$n), na.rm = TRUE)
  }
}
ep_perf      <- data.frame(tp=tp, fp=fp, fa=no_fa)
ep_perf

data.frame(mean_tp=mean(ep_perf$tp), mean_fa = mean(ep_perf$fa), mean_nofa= mean(ep_perf$fa))






scripts/epR_testscript.R

library(endpointR)

file      <- "C:/MHH Bleich/Papers/Schwabe App/Data/test/Validation_data.txt"
testdata  <- ep_load(file )

a         <- 1 #50
td        <- ep_select(testdata, a) #td        <- ep_select(testdata, "WZ12")

result    <- epR(td        = td,
                 org       = F,
                 wl        = 6,
                 SDwdth    = 2,
                 mad       = F,
                 cex       = 1.4,
                 cex.lab   = 1.2,
                 blind     = F,
                 ignupr    = FALSE,
                 ylim      = c(90,105), #c(300,350),
                 xlim      = c(0,28),
                 xlab      = "day")
result




i =15
td        = as.numeric(ep_select(gliodat, i))
org       = F
wl        = 6
SDwdth    = 2
mad       = F
cex       = 1.4
cex.lab   = 1.2
blind     = F
ignupr    = FALSE
ylim      = c(90,120) #c(300,350),
xlim      = c(0,28)
xlab      = "day"
ylab      = "MA"
cex.axis  = 1
cex.lab   = 1
ltype     = "b"
pch       = 19
dotcolor  = "black"
main      ="test"
uprcol    = "darkgreen"
lwrcol    = "magenta"


